axelor-stock/src/main/java/com/axelor/apps/stock/service/StockMoveLineServiceImpl.java

```diff
@@ -844,4 +847,47 @@ public BigDecimal computeNetMass(StockMoveLine stockMoveLine, Company company)
             .convertWithProduct(startUnit, endUnit, product.getNetMass(), product);
     return netMass;
   }
+
+  @Override
+  @Transactional
+  public void updateStockMoveLineFromTrackNumber(
+      StockMoveLine stockMoveLine, ArrayList<LinkedHashMap<String, Object>> trackingNumbers) {
+    boolean draft = true;
+    if (stockMoveLine.getStockMove().getStatusSelect() == StockMoveRepository.STATUS_PLANNED) {
+      draft = false;
+    }
+
+    for (LinkedHashMap<String, Object> trackingNumberItem : trackingNumbers) {
+      BigDecimal counter = new BigDecimal(trackingNumberItem.get("counter").toString());
+
+      TrackingNumber trackingNumber = new TrackingNumber();
+      trackingNumber.setCounter(counter);
+      trackingNumber.setTrackingNumberSeq(trackingNumberItem.get("trackingNumberSeq").toString());
+      if (trackingNumberItem.get("warrantyExpirationDate") != null) {
+        trackingNumber.setWarrantyExpirationDate(
+            LocalDate.parse(trackingNumberItem.get("warrantyExpirationDate").toString()));
+      }
+      if (trackingNumberItem.get("perishableExpirationDate") != null) {
+        trackingNumber.setPerishableExpirationDate(
+            LocalDate.parse(trackingNumberItem.get("perishableExpirationDate").toString()));
+      }
+      trackingNumber.setProduct(stockMoveLine.getProduct());
+      Beans.get(TrackingNumberRepository.class).save(trackingNumber);
+
+      StockMoveLine stMoveLine = new StockMoveLine();
```

Comment:
I think it's better to use JPA.Copy method here instead of new, because, else, we will loss some informations.